### ISSUE TITLE
RemoteSpringApplication add/remove class issue

### ISSUE THREAD

**Issue by paweln1986** [spring-projects/spring-boot#7379]
I am using `RemoteSpringApplication` and it is working correctly for modifying class content. I am not able to add or remove class. New classes are not visible by Spring context after successful reload. I have 

```
2016-11-14 08:00:12.746  INFO 9762 --- [   File Watcher] o.s.b.d.r.c.ClassPathChangeUploader      : Uploaded 1 class resource
2016-11-14 08:00:14.484  INFO 9762 --- [pool-1-thread-1] o.s.b.d.r.c.DelayedLiveReloadTrigger     : Remote server has changed, triggering LiveReload
``` 
but no new controllers are visible. Same for local development works out of the box.   
Similar with class removal but in this case we have got exception on server side.   

```
Caused by: java.io.FileNotFoundException: class path resource [com/example/TestRest1.class] cannot be opened because it does not exist
```

Full stack attached [here](https://gist.github.com/paweln1986/2cee4656c14c82089a7bb30f7cddc985)
Is it my miss-configuration or a bug?

[Sample project](https://github.com/paweln1986/RemoteSpringApplicationBug)
My IntellIJ configuration.
![test2](https://cloud.githubusercontent.com/assets/995565/20336776/1061a7d0-abcd-11e6-9a26-ec1985f9890b.png)
Simplest way to reproduce this issue:

1. java -jar test-0.0.1-SNAPSHOT.jar
2. Run RemoteSpringApplication from IntellIJ
3. Add new controller and compile whole project. RemoteSpringApplication indicate resources are updated but nothing was changed and new controller is not visible.
4. Remove HelloWorld controller. SpringBootApp crashes with exception.

**Comment by philwebb** [comment]
Sounds like it could be a bug. Are you able to share a sample project that shows the issue?

**Comment by paweln1986** [comment]
My question was updated.

**Comment by wilkinsona** [comment]
@paweln1986 Thanks for the update. I've reproduced the problem.

**Comment by wilkinsona** [comment]
There are two things going wrong here:

 `RestartServer` doesn't update the file system when the application's run as a jar as it requires a URL on the class path to point to a folder. We don't seen this problem with apps deployed to Cloud Foundry as they're automatically unpacked as part of the deployment. This behaviour can easily be mimicked by unzipping the jar and then running it:

```
$ mkdir unpacked
$ cd unpacked
$ unzip ../test-0.0.1-SNAPSHOT.jar
$  java -cp . org.springframework.boot.loader.JarLauncher
```

The second problem is that `RestartServer` is only capable of deleting or updating existing files. It cannot add new files:

```java
if (file.exists() && file.canWrite()) {
	if (classLoaderFile.getKind() == Kind.DELETED) {
		return file.delete();
	}
	FileCopyUtils.copy(classLoaderFile.getContents(), file);
	return true;
}
```

It's not immediately clear to me how to fix the second part of the problem. We run as an unpacked jar there are multiple folders on the class path and the restarter would have to know which one to use when adding new files to the application. We could hardcode it to use `BOOT-INF/classes` but that would couple DevTools to a particular layout. However, it then wouldn't work for an unpacked war where it should use `WEB-INF/classes`.

**Comment by wilkinsona** [comment]
[This branch](https://github.com/wilkinsona/spring-boot/tree/gh-7379) contains some changes that, I think, are a step in the right direction.

**Comment by oak-tree** [comment]
cool fix, quick question: which version of spring-boot contains these patches?

**Comment by snicoll** [comment]
@oak-tree are you browsing the mobile version of the site in any chance? There is a milestone right to the comment section on the desktop site. It's fixed as of 1.4.3 (not released yet)

**Comment by oak-tree** [comment]
I see it now, thanks!
